test_unit.yml, tests_integration_pytest.yml - support the collection converted format.

nhosoi opened the pull request:
tests_integration_pytest.yml

    To run the pytest on the target hosts via CI tests, integration_pytest_
    python3.yml copies entire network role to the target. To get the top
    directory, it uses a git command line. But the method cannot be used
    in the collection converted format since the converted roles and tests
    are not located in the git-cloned directory.
    
    In the collection converted format, the network tests directory has the
    copy of modules and module_utils. Taking advantage of the self-contained
    tests, we can just copy the files and directories in the tests and run
    test_integration_pytest successfully in the target host.

tests_unit.yml

    The unit test depends on the relative paths to library and module_utils,
    which are no longer available from the tests dir. (The tests directory
    in the collections CI is independent from the collections path.) To
    workaround it, the library and module_utils are copied to the tests
    directory and let the unit test additionally refer them.

richm commented on tests/integration/test_ethernet.py:
```diff
@@ -12,7 +12,11 @@
     import mock
 
-parentdir = os.path.normpath(os.path.join(os.path.dirname(__file__), "../../"))
+_is_role = os.path.exists(os.path.join(os.path.dirname(__file__), "../roles/linux-ystem-roles.network"))
```
linux-system-roles.network

nhosoi replied:
*shame shame...* Now I wonder why my tests (especially, in the legacy role env) successfully ran...

richm commented on tests/playbooks/integration_pytest_python3.yml:
```diff
@@ -42,38 +42,62 @@
+            rsync_opts:
+              - "--exclude=.pyc"
+              - "--exclude=__pycache__"
+          when: false
```
when false?

richm replied:
ah, I see

richm commented on tests/playbooks/integration_pytest_python3.yml:
```diff
@@ -42,38 +42,62 @@
+
+        # TODO: using tar and copying the file is a workaround for the synchronize
+        # module that does not work in test-harness. Related issue:
+        # https://github.com/linux-system-roles/test-harness/issues/102
```
This has been fixed - but I guess that should be addressed in a separate commit

richm commented on tests/tests_unit.yml:
```diff
@@ -29,15 +29,24 @@
         - unit/test_network_connections.py
         - ../module_utils/network_lsr
+        - module_utils/network_lsr
+        - modules/network_connections.py
```
what is the path for these?  That is, in which directory is this task assuming it is being run from?

nhosoi replied:
In the `tests` directory. There are symlinks `modules` targeting `../library` and `module_utils` targeting `../module_utils`. When converting to the collections format, the symlinks are replaced with the real dirs and files to avoid having symlinks in the collections format.

nhosoi commented:
[citest]

cathay4t commented:
@nhosoi Could you rebase which will include the fixes for deactivatin/down a connection?

cathay4t reviewed (dismissed):
LGTM.

nhosoi commented:
Thank you, @richm and @cathay4t.

Rebased and squashed the commit for fixing the issues found in the review into the previous one. (I'm keeping one commit per each test playbook.)

Once the CI tests finish, could you please merge since I do not have the write access. Thanks, in advance.

nhosoi commented:
[citest bad]

cathay4t commented:
@nhosoi  There are some error regarding `ansible_collections.fedora.system_roles.plugins.module_utils.network_lsr.nm.provider.__spec__` on RHEL 8.
Should it been fixed in separate PR?

nhosoi commented:
[citest bad]

nhosoi commented:
> @nhosoi There are some error regarding `ansible_collections.fedora.system_roles.plugins.module_utils.network_lsr.nm.provider.__spec__` on RHEL 8.
> Should it been fixed in separate PR?

@cathay4t, Sorry for my ignorance. Do you happen to know what is causing this error and how to fix it? If I run the same CI test script locally on my laptop, the tests pass without the error...
```
<<hostname>>.localdomain: linux-system-roles/network: pull #319 (a4d19a4) on rhel-8
==============================================================================

Use cached rhel-8.qcow2
+ git init --quiet /tmp/tmp41ogb26u
+ GIT_TERMINAL_PROMPT=0 git -C /tmp/tmp41ogb26u fetch --quiet --depth=1 https://github.com/linux-system-roles/network pull/319/head
+ git -C /tmp/tmp41ogb26u checkout --quiet FETCH_HEAD
Run ansible-playbook with environment variable ANSIBLE_COLLECTIONS_PATHS=/tmp/tmpzlexpj8j:~/.ansible/collections:/usr/share/ansible/collections
Testing /tmp/tmp4h4gwne7/tests/network/tests_802_1x_nm.yml...SUCCESS
Testing /tmp/tmp4h4gwne7/tests/network/tests_802_1x_updated_nm.yml...SUCCESS
<<snip>>
```

nhosoi commented:
[citest bad]

nhosoi commented:
[citest bad]

nhosoi commented:
[citest bad]

nhosoi commented:
[citest bad]

nhosoi commented:
[citest bad]

nhosoi commented:
> @nhosoi There are some error regarding `ansible_collections.fedora.system_roles.plugins.module_utils.network_lsr.nm.provider.__spec__` on RHEL 8.
> Should it been fixed in separate PR?

@cathay4t, @richm, I'm afraid we might have hit a known issue in the collections which freeipa had to go through (or similar).
https://github.com/freeipa/ansible-freeipa/issues/230
https://github.com/ansible/ansible/issues/68361
It looks ansible-freeipa could workaround the problem with this pr.
https://github.com/freeipa/ansible-freeipa/pull/233
@t-woerner, do you think linux-system-roles share the same issue you solved with the ansible-freeipa pr233? If so, could you please give us your advice how to workaround it? Collections CI tests are failing if the linux-system-roles have the custom module_utils, which are certificate, network and storage...
```
TASK [fedora.system_roles.network : Configure networking connection profiles] ***
task path: /tmp/tmpttpq1nm6/ansible_collections/fedora/system_roles/roles/network/tasks/main.yml:80
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: ValueError: ansible_collections.fedora.system_roles.plugins.module_utils.network_lsr.nm.provider.__spec__ is None
fatal: [/cache/centos-8.qcow2]: FAILED! => {"msg": "Unexpected failure during module execution.", "stdout": ""}
```

nhosoi commented:
[citest bad]

nhosoi commented:
[citest bad]

nhosoi commented:
[citest bad]

nhosoi commented:
[citest bad]

nhosoi commented:
[citest bad]

richm commented on tests/playbooks/integration_pytest_python3.yml:
```diff
@@ -42,38 +42,66 @@
-             --exclude "*.pyc" --exclude "__pycache__" --exclude testrepo.tar
-             -C {{ git_top_directory.stdout }} .'
-      delegate_to: localhost
+    - name: Determine legacy role format or collection format
```
Might I suggest something like this instead?  https://gist.github.com/richm/f3338790eb48b6a3f0963f908a2f4a03

That is
* assume that if running as a role, the tests directory will be `../../tests`, the modules directory will be `../../library`, and the module_utils directory will be `../../module_utils`
* otherwise, look in the various ansible paths, including the collection paths, for the modules and module_utils

nhosoi replied:
Hi @richm, could you please review this commit?
https://github.com/linux-system-roles/network/pull/319/commits/ce7b70dd2f93cc3283317149caef7892d9e0d6df

I had to tweak somehow to make it work on the control node. Hope it's ok...

Good news is staging tests with ansible 2.9 successfully run now!

nhosoi commented:
[citest bad]

nhosoi commented:
[citest bad]

nhosoi commented:
[citest bad]

nhosoi commented:
[citest bad]

nhosoi commented:
[citest bad]

nhosoi commented:
[citest bad]

nhosoi commented:
[citest bad]

nhosoi commented:
[citest bad]

nhosoi commented:
[citest bad]

nhosoi commented:
[citest bad]

nhosoi commented:
[citest bad]

nhosoi commented:
[citest bad]

nhosoi commented:
[citest bad]

nhosoi commented:
[citest bad]

nhosoi commented:
[citest bad]

nhosoi commented:
[citest]

nhosoi commented:
[citest bad]

nhosoi commented:
[citest bad]

nhosoi commented:
[citest]

nhosoi commented:
[citest]

nhosoi commented:
[citest bad]

nhosoi commented:
[citest]

nhosoi commented:
[citest]

nhosoi commented:
[citest pending]

richm commented:
Does tests_unit.yml work with collections?

nhosoi commented:
> Does tests_unit.yml work with collections?

The network tests dir has modules and module_utils symlinks targeting library and module_utils, respectively.
```
$ ls -l modules module_utils
lrwxrwxrwx. 1 nhosoi nhosoi 11 May  2  2020 modules -> ../library//
lrwxrwxrwx. 1 nhosoi nhosoi 16 May  2  2020 module_utils -> ../module_utils//
```
The symlinks are replaced with the real files and directories by lsr_role2collection in the conversion process, which make the unit tests pass. That's said, the unit test in the collections format does not test the "format", but tests the functionality in the same way as the legacy role unit test does...

richm commented:
> The symlinks are replaced with the real files and directories by lsr_role2collection in the conversion process, which make the unit tests pass

I think lsr_role2collection should not touch these symlinks, and should not copy/convert symlinks at all.

nhosoi commented:
> > The symlinks are replaced with the real files and directories by lsr_role2collection in the conversion process, which make the unit tests pass
> 
> I think lsr_role2collection should not touch these symlinks, and should not copy/convert symlinks at all.

I thought the collections format is not supposed to have symlinks... That was the motivation to convert the symlinks to the real files and directories... In this case, if the tool does not touch the symlinks, the links will be broken in the converted format. Do we want to remove such symlinks and make the tests work not depending on them?

nhosoi commented:
This PR is closed in favour of https://github.com/linux-system-roles/network/pull/328.
